### PR TITLE
borrow and bookcase

### PR DESCRIPTION
**Pull request by doanhvv**

### CODE REVIEW

**Comment by doanhvv** [comment]
ready

File: app/Http/Controllers/UserController.php [hunk 1]
```diff
@@ -40,4 +42,66 @@ public function favorite()
...
+    {
+        $book = Book::findorFail($id);
+
+        return view('frontend.user.borrow', compact('book'));
+    }
+
+    public function borrow(Request $request, $id)
+    {
+        if(!$this->checkAvailableBook($id)){
+            return redirect()->back()->with('error', 'Rất tiếc, kho đã hết sách, hẹn gặp lại.');
```
Comment (tranvanmy) [inline app/Http/Controllers/UserController.php]: Please use i18n

File: app/Http/Controllers/UserController.php [hunk 2]
```diff
@@ -40,4 +42,66 @@ public function favorite()
...
+                'start_time' => $request->starttime,
+                'end_time' => $request->endtime,
+                'status' => 1
+            ]);
+
+            $book = Book::findorFail($id);
+            $book->quantity = $book->quantity - 1;
+            $book->update();
+
+            return redirect()->route('bookcase')->with('success', 'Bạn đã đăng kí mượn sách thành công, sẽ mất một khoảng thời gian để đơn đăng kí được duyệt.');
```
Comment (tranvanmy) [inline app/Http/Controllers/UserController.php]: Please use i18n

File: app/Http/Controllers/UserController.php [hunk 3]
```diff
@@ -40,4 +42,66 @@ public function favorite()
...
+            ]);
+
+            $book = Book::findorFail($id);
+            $book->quantity = $book->quantity - 1;
+            $book->update();
+
+            return redirect()->route('bookcase')->with('success', 'Bạn đã đăng kí mượn sách thành công, sẽ mất một khoảng thời gian để đơn đăng kí được duyệt.'); 
+        }
+
+        return redirect()->back()->with('error', 'Ngày mượn sách phải nhỏ hơn hoặc bằng ngày trả sách.');
```
Comment (tranvanmy) [inline app/Http/Controllers/UserController.php]: Please use i18n

File: app/Http/Controllers/UserController.php [hunk 4]
```diff
@@ -40,4 +42,66 @@ public function favorite()
...
+        $borrow = Borrow::findOrFail($id);
+        if($borrow->user_id == auth()->id()){
+            $borrow->update(['status' => config('customs.book.status.cancel')]);
+
+            return redirect()->back();
+        }
+
+        return abort('403');
+    }
+}
```
Comment (tranvanmy) [inline app/Http/Controllers/UserController.php]: convention

File: resources/lang/en/enum.php [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+return [
+    'waiting' => 'Chờ duyệt',
+    'approved' => 'Đã duyệt',
+    'out_of_date' => 'Quá hạn',
+    'returned' => 'Đã trả sách',
+    'unknown' => 'Không xác định',
+    'cancel' => 'Bị hủy',
+    'not_approved' => 'Không được duyệt',
+];
```
Comment (tranvanmy) [inline resources/lang/en/enum.php]: convention

File: resources/views/frontend/user/bookcase.blade.php [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+@section('header')
+    
+@endsection
+
+@section('content')
+<div class="container">
+    <div class="row">
+        <div class="col-md-12">
+            <div class="image-book">
+                <h1>Tủ sách</h1>
```
Comment (tranvanmy) [inline resources/views/frontend/user/bookcase.blade.php]: Please use i18n

File: resources/views/frontend/user/bookcase.blade.php [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+</div>
+<div class="book bg-skew bg-skew-light">
+    <div class="container">
+        <div class="row">
+            <div class="col-md-12">
+                @if (session()->has('success'))
+                    <div class="alert alert-success" role="alert">{{ session()->get('success') }}</div>
+                @endif
+                
+                <p class="text-warning">Vui lòng đọc quy định mượn trả sách của chúng tôi <a href="">tại đây.</a></p>
```
Comment (tranvanmy) [inline resources/views/frontend/user/bookcase.blade.php]: Please use i18n

File: resources/views/frontend/user/bookcase.blade.php [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+                    <table class="table table-hover">
+                        <thead>
+                            <tr>
+                                <th>STT</th>
+                                <th>Ảnh</th>
+                                <th>Tên sách</th>
+                                <th>Ngày mượn</th>
+                                <th>Hạn trả</th>
+                                <th>Trạng thái</th>
+                                <th>#</th>
```
Comment (tranvanmy) [inline resources/views/frontend/user/bookcase.blade.php:40]: Please use i18n

File: resources/views/frontend/user/bookcase.blade.php [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+                                        @elseif($borrow->status == config('customs.book.status.returned'))
+                                            <p class="text-default">{{ trans('enum.returned') }}</p>
+                                        @else
+                                            <p class="text-default">{{ trans('enum.unknown') }}</p>
+                                        @endif
+                                    </td>
+                                    <td>
+                                        @if($borrow->status == config('customs.book.status.waiting'))
+                                            <a href="{{ route('book.cancel', $borrow->id) }}" onclick="return confirm('Bạn không muốn mượn cuốn này nữa ?');">Hủy</a>
+                                        @endif
```
Comment (tranvanmy) [inline resources/views/frontend/user/bookcase.blade.php:73]: Please use i18n

File: resources/views/frontend/user/borrow.blade.php [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+@section('header')
+    
+@endsection
+
+@section('content')
+<div class="container">
+    <div class="row">
+        <div class="col-md-12">
+            <div class="image-book">
+                <h1>Mượn sách</h1>
```
Comment (tranvanmy) [inline resources/views/frontend/user/borrow.blade.php]: Please use i18n

File: resources/views/frontend/user/borrow.blade.php [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+            </div>
+        </div>
+    </div>
+</div>
+<div class="book bg-skew bg-skew-light">
+    <div class="container">
+        <div class="row">
+            <div class="col-md-6">
+                <h3>Tên sách: {{ $book->name }}</h3>
```
Comment (tranvanmy) [inline resources/views/frontend/user/borrow.blade.php]: Please use i18n

File: resources/views/frontend/user/borrow.blade.php [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+                    @csrf
+
+                    <div class="form-group">
+                        <label for="">Ngày mượn sách:</label>
+                        <input name="starttime" type="date" class="form-control" required>
+                    </div>
+                
+                    <div class="form-group">
+                        <label for="">Ngày trả sách:</label>
+                        <input name="endtime" type="date" class="form-control" required>
```
Comment (tranvanmy) [inline resources/views/frontend/user/borrow.blade.php:41]: Please use i18n

File: routes/web.php [hunk 13]
```diff
@@ -32,6 +32,10 @@
     Route::get('user/profile/changepassword', 'UserController@showChangePasswordForm')->name('changepassword');
     Route::post('user/profile/changepassword', 'UserController@changePassword')->name('changepassword');
     Route::get('user/favorite', 'UserController@favorite')->name('favorite');
+    Route::get('user/borrow/{id?}', 'UserController@showBorrowForm')->name('borrow');
+    Route::post('user/borrow/{id?}', 'UserController@borrow')->name('borrow');
```
Comment (tranvanmy) [inline routes/web.php]: why use "?"  in route ?????

File: routes/web.php [hunk 14]
```diff
@@ -45,3 +49,7 @@
     Route::resource('authors', 'AuthorController');
     Route::resource('books', 'BookController');
 });
+
+Route::get('test', function(){
+    echo now();
+});
```
Comment (tranvanmy) [inline routes/web.php]: Please remove if not use

**Review by tranvanmy — CHANGES_REQUESTED** [review]
Please check comment

File: app/Http/Controllers/UserController.php [hunk 15]
```diff
@@ -40,4 +42,66 @@ public function favorite()
...
+    }
+
+    public function borrow(Request $request, $id)
+    {
+        if(!$this->checkAvailableBook($id)){
+            return redirect()->back()->with('error', 'Rất tiếc, kho đã hết sách, hẹn gặp lại.');
+        }
+
+        if(strtotime($request->starttime) <= strtotime($request->endtime)){
+            $borrow = Borrow::create([
```
Comment (tranvanmy) [inline app/Http/Controllers/UserController.php]: should use 'firstOrCreate'

File: app/Http/Controllers/UserController.php [hunk 16]
```diff
@@ -40,4 +42,66 @@ public function favorite()
...
+                'book_id' => $id,
+                'user_id' => Auth::id(),
+                'start_time' => $request->starttime,
+                'end_time' => $request->endtime,
+                'status' => 1
+            ]);
+
+            $book = Book::findorFail($id);
+            $book->quantity = $book->quantity - 1;
+            $book->update();
```
Comment (tranvanmy) [inline app/Http/Controllers/UserController.php]: use try { } catch and transaction for function

File: app/Http/Controllers/UserController.php [hunk 17]
```diff
@@ -40,4 +42,66 @@ public function favorite()
...
+        return redirect()->back()->with('error', 'Ngày mượn sách phải nhỏ hơn hoặc bằng ngày trả sách.');
+    }
+
+    public function checkAvailableBook($id)
+    {
+        $book = Book::findorFail($id);
+        if($book->quantity > 0){
+            return TRUE;
+        }
+        return FALSE;
```
Comment (tranvanmy) [inline app/Http/Controllers/UserController.php]: please format this

File: app/Http/Controllers/UserController.php [hunk 18]
```diff
@@ -40,4 +42,66 @@ public function favorite()
...
+        $borrows = Borrow::with('book')->where('user_id', auth()->id())
+        ->orderBy('id', 'desc')->paginate(10);
+
+        return view('frontend.user.bookcase', compact('borrows'));
+    }
+
+    public function cancelBorrowBook($id)
+    {
+        $borrow = Borrow::findOrFail($id);
+        if($borrow->user_id == auth()->id()){
```
Comment (tranvanmy) [inline app/Http/Controllers/UserController.php:134]: convention

File: resources/views/frontend/book/detail.blade.php [hunk 19]
```diff
@@ -38,7 +38,11 @@
                     </p>
                     
                     <div class="book-button">
-                        <a class="btn btn-pill btn-primary" href="#" role="button">Mượn sách</a>
+                        @if (auth()->check())
+                            <a class="btn btn-pill btn-primary" href="{{ route('borrow', $book->id) }}" role="button">Mượn sách</a>
+                        @else
```
Comment (tranvanmy) [inline resources/views/frontend/book/detail.blade.php:42]: Please use i18n

File: routes/web.php (hunk 13 above)
Comment (tranvanmy) [inline routes/web.php]: Why use "?" in route ??

**Comment by doanhvv** [comment]
ready

**Review by tranvanmy — APPROVED** [review]

File: app/Http/Controllers/UserController.php [hunk 20]
```diff
@@ -40,4 +45,103 @@ public function favorite()
...
+    {
+        $book = Book::findorFail($id);
+
+        return view('frontend.user.borrow', compact('book'));
+    }
+
+    public function borrow(Request $request, $id)
+    {
+        if($this->checkBorrowBook($id)){
+
```
Comment (tranvanmy) [inline app/Http/Controllers/UserController.php]: bỏ khoảng trắng đi nhé,

**Pull request merged**